### ISSUE TITLE
Feature Request: Resources for modification of child objects/tables

### ISSUE THREAD

**Issue by poroping** [fortinetdev/terraform-provider-fortios#127]
Specifically a resource to interact with the BGP neighbors table directly. I am writing a reusable module to setup a site-site VPN w/ BGP peering with the public cloud providers. However you can only modify the neighbors table using the fortios_router_bgp resource which would then override any existing config.

Being able to modify child tables directly would open up a lot more functionality and ability to write more modular modules.

Would need to specify that if you were using the 'parent' resource that the 'child' resources would be over written however I've seen similar caveats with the Azure resources so I don't believe this would be breaking any Terraform rules.

It's not documented but you can modify these tables directly using POST, PUT and DEL. For example my work around:

`resource "fortios_json_generic_api" "post_test1" {
  path   = "/api/v2/cmdb/router/bgp/neighbor"
  method = "POST"
  json   = <<EOF
{
  "ip": "4.4.4.4",
  "remote-as": "65003"
}
EOF
}`

Obviously this doesn't correctly store the state and is overwritten by the 'proper' resource if used which isn't ideal.

**Comment by frankshen01** [comment]
Hi @poroping, Thanks for the feature request. I think this makes sense. If I understand correctly, your requirements can be separated into two parts:
1. When configuring BGP, we can specify whether to override neighbors table
2. Add a new resource, we can directly operate neighbors resource, so we can avoid using json generic api which lacks the ability to support state storage. 

Is this right?

**Comment by poroping** [comment]
1. Yes, for example in my situation if we already have neighbor(s)
configured, whether already in state or not, the fortios_router_bgp
resource will delete these neighbors and add the configured neighbor. So
being able to specify to 'merge' instead of 'replace' would be great.

2. If you can implement 1. Then this might not be necessary. It just might
be cleaner and/or easier to implement as separate resources.

I believe this behavior applies to all the child objects in BGP and I'd
assume the other routing protocols.

I'll provide some examples when I can over the holiday break.

**Comment by poroping** [comment]
BGP config before applying resource (doesn't matter if this is already managed by terraform or not, will still defined blocks)

```
config router bgp
    set as 65002
    config neighbor
        edit "3.3.3.3"
            set remote-as 54455
        next
        edit "4.4.4.4"
            set remote-as 65003
        next
    end
    config network
        edit 1
            set prefix 10.0.0.0 255.255.255.0
        next
    end
    config redistribute "connected"
    end
    config redistribute "rip"
    end
    config redistribute "ospf"
    end
    config redistribute "static"
    end
    config redistribute "isis"
    end
    config redistribute6 "connected"
    end
    config redistribute6 "rip"
    end
    config redistribute6 "ospf"
    end
    config redistribute6 "static"
    end
    config redistribute6 "isis"
    end
end
```

Resource applied (AS is required)

```
resource "fortios_router_bgp" "fgt_bgp" {
  as = 65000
  neighbor {
    ip                   = "10.10.10.10"
    remote_as            = 65001
    soft_reconfiguration = "enable"
  }
}
```

Fortigate config after terraform apply, previously configured neighbors removed and new one added, networks not touched as network block not defined.

```
config router bgp
    set as 65002
    config neighbor
        edit "10.10.10.10"
            set soft-reconfiguration enable
            set remote-as 65001
        next
    end
    config network
        edit 1
            set prefix 10.0.0.0 255.255.255.0
        next
    end
    config redistribute "connected"
    end
    config redistribute "rip"
    end
    config redistribute "ospf"
    end
    config redistribute "static"
    end
    config redistribute "isis"
    end
    config redistribute6 "connected"
    end
    config redistribute6 "rip"
    end
    config redistribute6 "ospf"
    end
    config redistribute6 "static"
    end
    config redistribute6 "isis"
    end
end
```

terraform destroy doesn't seem to do anything but I assume this can be a separate issue:

```
fgt (root) # diagnose debug enable

fgt (root) # diagnose debug application httpsd -1
Debug messages will be on for 30 minutes.

fgt (root) # diagnose debug cli 8
Debug messages will be on for 30 minutes.

fortios_router_bgp.fgt_bgp: Destroying... [id=RouterBgp]
fortios_router_bgp.fgt_bgp: Destruction complete after 0s

fgt (root) # config router bgp

fgt (bgp) # show
path=router, objname=bgp, tablename=(null), size=296
config router bgp
    set as 65000
    config neighbor
        edit "10.10.10.10"
            set soft-reconfiguration enable
            set remote-as 65001
        next
    end
    config network
        edit 1
            set prefix 10.0.0.0 255.255.255.0
        next
    end
    config redistribute "connected"
    end
    config redistribute "rip"
    end
    config redistribute "ospf"
    end
    config redistribute "static"
    end
    config redistribute "isis"
    end
    config redistribute6 "connected"
    end
    config redistribute6 "rip"
    end
    config redistribute6 "ospf"
    end
    config redistribute6 "static"
    end
    config redistribute6 "isis"
    end
end
```

Tested with 1.6.18

Ideally I'd like to be able to do something like:

```
resource "fortios_router_bgp" "fgt_bgp_settings" {
  as = 65000
}

resource "fortios_router_bgp_neighbor" "fgt_bgp_neighbor" {
  ip                   = "10.10.10.10"
  remote_as            = 65001
  soft_reconfiguration = "enable"
}
```

This would then allow users to add/remove neighbors/networks etc. without having to define the entire config in a single resource, each network/neighbor could be it's own resource. It would need to be noted that defining a neighbor block in the 'parent' resource would overwrite existing neighbors (which is the current behavior).

Alternatively a 'merge' key could be added to the fortios_router_bgp resource to specify the config is to be merged with the current config, however I feel the additional resources would be cleaner, easier to read and probably implement as they should all have their own endpoints in the API.

Let me know what you think.

**Comment by frankshen01** [comment]
Hi @poroping, thanks for this thoughtful discussion and analysis. I have created a new branch `test-tf-nb` to help discuss this topic.

**1 Branch Info**

Branch link: https://github.com/fortinetdev/terraform-provider-fortios/commits/test-tf-nb

New features:
- new resource `fortios_routerbgp_neighbor`
- new data source `data_source_router_bgp`
- `force_destroy` for clearing sub tables that have not been set
- clean sub-table for destroy

**2 Behavior of BGP API**

Assume the following settings on the firewall:

```
# config router bgp

myfirewall (bgp) # show
config router bgp
    set as 65000
    config neighbor
        edit "41.10.10.10"
            set soft-reconfiguration enable
            set remote-as 65003
        next
        edit "51.1.1.1"
            set soft-reconfiguration enable
            set remote-as 65002
        next
    end
    config network
        edit 1
            set prefix 33.0.0.0 255.255.255.0
        next
        edit 2
            set prefix 133.0.0.0 255.255.255.0
        next
    end
    config redistribute "connected"
    end
    config redistribute "rip"
    end
    config redistribute "ospf"
    end
    config redistribute "static"
    end
    config redistribute "isis"
    end
    config redistribute6 "connected"
    end
    config redistribute6 "rip"
    end
    config redistribute6 "ospf"
    end
    config redistribute6 "static"
    end
    config redistribute6 "isis"
    end
end

```
Then we configure new neighbors and networks through curl:

```
curl --location --request PUT 'https://192.168.52.177/api/v2/cmdb/router/bgp?access_token=zg3H1x4gcykx6b9h5b44j75hxd9dfg' \
--header 'Content-Type: application/json' \
--data-raw '{
    "as": 65002,
    "neighbor": [
        {
            "ip": "20.10.10.10",
            "remote-as": 65001,
            "soft_reconfiguration ": "enable"
        }
    ],
    "network": [
        {
            "prefix": "33.0.0.0 255.255.255.0",
            "backdoor": "disable",
            "route-map": ""
        }
    ]
}'

```
Let's see the result:
```
myfirewall (bgp) # show
config router bgp
    set as 65000
    config neighbor
        edit "20.10.10.10"
            set remote-as 65001
        next
    end
    config network
        edit 1
            set prefix 122.0.0.0 255.255.255.0
        next
    end
    config redistribute "connected"
    end
    config redistribute "rip"
    end
    config redistribute "ospf"
    end
    config redistribute "static"
    end
    config redistribute "isis"
    end
    config redistribute6 "connected"
    end
    config redistribute6 "rip"
    end
    config redistribute6 "ospf"
    end
    config redistribute6 "static"
    end
    config redistribute6 "isis"
    end
end
```
It can be seen that any new sub-table will overwrite all the old values of the sub-table including network. If the sub-table is not set, the old values will remain there by default, but if the sub-table is assigned a null value in the API, for example [] , the sub-table will be emptied.

The designers of FortiAPI may have designed this based on a principle that they believe is reasonable, terraform also followed the API design. In addition, because the API only provides GET and PUT API method, and no Delete method is provided, so if we want to use the API to empty some sub-table, assign a null value to the sub-table to empty the sub-table. It has been supported in the branch.

**3 About the operation of the API sub-table:**

Many FortiAPI sub-tables have such interfaces, including GET, PUT, POST, DELETE, etc. E.g:

- api/v2/cmdb/firewall.service/group/s111/member
- api/v2/cmdb/system/api-user/1/trusthost

The number of sub-tables and nested sub-tables contained in more than 400 API resources is huge.

**4 Problems caused by mixed operation of main table resources fortios_router_bgp and sub-table resources fortios_routerbgp_neighbor**:

First let us see the following example (with the provider compiled by the test-tf-nb branch):
```
provider "fortios" {
  hostname = "192.168.52.177"
  token    = "zg3H1x4gcykx6b9h5b44j75hxd9dfg"
  insecure = "true"
}

resource "fortios_router_bgp" "fgt_bgp" {
  as = 65000

  neighbor {
    ip                   = "121.1.1.1"
    remote_as            = 65003
    soft_reconfiguration = "enable"
  }
  neighbor {
    ip                   = "122.1.1.1"
    remote_as            = 65003
    soft_reconfiguration = "enable"
  }

}

resource "fortios_routerbgp_neighbor" "s1" {
    ip                   = "123.1.1.1"
    remote_as            = 65003
    soft_reconfiguration = "enable"

  depends_on = [
    fortios_router_bgp.fgt_bgp,
  ]
}
```


```
# terraform apply
  ----

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # fortios_router_bgp.fgt_bgp will be created
  + resource "fortios_router_bgp" "fgt_bgp" {
      + as                                 = 65000
      + force_destroy                      = false

      + neighbor {
          .................
          + ip                            = "121.1.1.1"
          + remote_as                     = 65003
          + soft_reconfiguration          = "enable"
          + filter_list_in6               = (known after apply)
          + filter_list_out               = (known after apply)
          + filter_list_out6              = (known after apply)
          + holdtime_timer                = (known after apply)
          + interface                     = (known after apply)
          .................
        }
      + neighbor {
          .................
          + ip                            = "122.1.1.1"
          + remote_as                     = 65003
          + soft_reconfiguration          = "enable"
          + filter_list_in6               = (known after apply)
          + filter_list_out               = (known after apply)
          + filter_list_out6              = (known after apply)
          + holdtime_timer                = (known after apply)
          + interface                     = (known after apply)
          .................
        }
    }

  # fortios_routerbgp_neighbor.s1 will be created
  + resource "fortios_routerbgp_neighbor" "s1" {
      + ip                            = "123.1.1.1"
      + remote_as                     = 65003
      + soft_reconfiguration          = "enable"
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

fortios_router_bgp.fgt_bgp: Creating...
fortios_router_bgp.fgt_bgp: Creation complete after 0s [id=RouterBgp]
fortios_routerbgp_neighbor.s1: Creating...
fortios_routerbgp_neighbor.s1: Creation complete after 0s [id=123.1.1.1]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```
See the result:

```
config router bgp
    set as 65000
    config neighbor
        edit "121.1.1.1"
            set soft-reconfiguration enable
            set remote-as 65003
        next
        edit "122.1.1.1"
            set soft-reconfiguration enable
            set remote-as 65003
        next
        edit "123.1.1.1"
            set soft-reconfiguration enable
            set remote-as 65003
        next
    end
    config redistribute "connected"
    end
    config redistribute "rip"
    end
    config redistribute "ospf"
    end
    config redistribute "static"
    end
    config redistribute "isis"
    end
    config redistribute6 "connected"
    end
    config redistribute6 "rip"
    end
    config redistribute6 "ospf"
    end
    config redistribute6 "static"
    end
    config redistribute6 "isis"
    end
end
```
We can find that the neighbor items created by fortios_router_bgp and fortios_routerbgp_neighbor are successfully sent to the firewall. But execute terraform plan, terraform will prompt that the status is inconsistent, as shown below:

```
# terraform plan
  ----
fortios_router_bgp.fgt_bgp: Refreshing state... [id=RouterBgp]
fortios_routerbgp_neighbor.s1: Refreshing state... [id=123.1.1.1]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # fortios_router_bgp.fgt_bgp will be updated in-place
  ~ resource "fortios_router_bgp" "fgt_bgp" {
        id                                 = "RouterBgp"
        # (42 unchanged attributes hidden)

      - neighbor {
          .................
          - ip                            = "123.1.1.1" -> null
          - keep_alive_timer              = 4294967295 -> null
          - link_down_failover            = "disable" -> null
          - local_as                      = 0 -> null
          - local_as_no_prepend           = "disable" -> null
          - local_as_replace_as           = "disable" -> null
          - maximum_prefix                = 0 -> null
          - maximum_prefix6               = 0 -> null
          - maximum_prefix_threshold      = 75 -> null
          - maximum_prefix_threshold6     = 75 -> null
          .................
        }
        # (2 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.

```
Because the neighbor items that fortios_router_bgp reads from the device also include the neighbor items created by fortios_routerbgp_neighbor, not just neighbor items created by itself, so terraform plan will always show inconsistent states.

**5 Merge: Sends all dynamic neighbor items to a BGP module for processing**

Regarding the implementation of MERGE feature, the BGP API doesn't support it. if we implement it by sub-table API, we will have to give up the operation of the top API on the sub-table, otherwise it will cause the problem mentioned in 4. Since the number of the sub-tables plus nested sub-tables contained in all more than 400 FortiAPI resources is surprisingly huge, if all resources are developed based on this method, and abandon the mechanism already implemented by Top API, I think this may not be the best choice, in addition, I don't want too many sub-table resources to dazzle ordinary users. Secondly, for terraform, one of the core challenges of infrastructure as code is keeping an up-to-date record of all deployed infrastructure and their properties. I think this is the main difference from configuration management tools. Terraform figures out how to get to that end state, Terraform will also be aware of any state it created in the past. Some configuration tools does not fully capture the state of the infrastructure. Terraform must store state about your managed infrastructure and configuration. This state is used by Terraform to map real world resources to the configuration. Therefore, in a certain resource, if its sub-table has been configured, then the sub-table attributes should be clearly known by the resource, and only appending to the sub-table means that the sub-table item before append is not controlled by the the resource, this is not conducive to terraform's state maintenance mechanism. The solution I can find (I believe you should have known it already), for example:

(1) Import the existing configuration on the device into terraform by importing in advance, so that terraform can keep awake for all sub-table configurations of the resource;

(2) Sends all neighbor items to a BGP module for processing. Example:

```
# tree
├── bgp2
│   ├── main.tf
│   └── versions.tf
├── maintst.tf
└── versions.tf
```

```
# cat maintst.tf
provider "fortios" {
  hostname = "192.168.52.177"
  token    = "zg3H1x4gcykx6b9h5b44j75hxd9dfg"
  insecure = "true"
}

module "bgp2" {
  source = "./bgp2"

  new_nblist_interface1 = [
    {
      ip                   = "221.11.1.1"
      remote_as            = 65003
      soft_reconfiguration = "enable"
    },
    {
      ip                   = "222.11.1.1"
      remote_as            = 65003
      soft_reconfiguration = "enable"
    }
  ]

  new_nblist_interface2 = [
    {
      ip                   = "31.11.1.1"
      remote_as            = 65003
      soft_reconfiguration = "enable"
    },
    {
      ip                   = "32.11.1.1"
      remote_as            = 65003
      soft_reconfiguration = "enable"
    },
    {
      ip                   = "33.11.1.1"
      remote_as            = 65003
      soft_reconfiguration = "enable"
    }
  ]
}


# cat bgp2/main.tf
variable "new_nblist_interface1" {
  type = list(object({
    ip = string
    remote_as = number
    soft_reconfiguration = string
  }))

  # default = []
}

variable "new_nblist_interface2" {
  type = list(object({
    ip = string
    remote_as = number
    soft_reconfiguration = string
  }))

  # default = []
}

resource "fortios_router_bgp" "fgt_bgp" {
  as = 65000

  dynamic "neighbor" {
    for_each = var.new_nblist_interface1
    content {
      ip                   = neighbor.value["ip"]
      remote_as            = neighbor.value["remote_as"]
      soft_reconfiguration = neighbor.value["soft_reconfiguration"]
    }
  }

  dynamic "neighbor" {
    for_each = var.new_nblist_interface2
    content {
      ip                   = neighbor.value["ip"]
      remote_as            = neighbor.value["remote_as"]
      soft_reconfiguration = neighbor.value["soft_reconfiguration"]
    }
  }
}

```

So that terraform can track the status of all sub-table neighbors' items.

**6 Destroy/clean problem**

For the destroy bug, it has been fixed in the branch and will be merged into the main version later, including related resources with the problem. Regarding whether the content of the unconfigured sub-table needs to be retained after apply, a new parameter force_destroy is added to handle this problem for user.

**My suggestions**

Based on the above experimental results and the current situation, I do not recommend the use of sub-table resources. Of course, if you think the neighbor sub-table resources need to be kept, I will merge it to master and release it after QA testing, but I plan to mark it as 'Debug' to prevent other users from using it, otherwise when the problem in 4 occurs, it will cause confusion to users.

Thank you!

**Comment by poroping** [comment]
Hey @frankshen01 thanks for the quick reply and the detailed and thorough answer.

I was thinking that the usage would be more along the lines of the Azure VNET and subnet resources: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/virtual_network
https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/subnet

You can define the subnets (our sub-table) in this top level resource, however a specific subnet resource also exists. Their documentation is clear that if you are going to use the subnet resource you cannot also define subnets within the VNET resource.

In this case the fortios_router_bgp resource could continue to operate much like it does today, where the sub tables are not modified unless the sub table block is defined. However I believe you would need to add functionality to purge the sub table if an empty block is defined. We could then clearly document that sub table resources exist but that you cannot use the relevant blocks in the top level resource as this would cause conflicts, as you noted in point 4.

**Example usage:**

Define BGP settings, don't declare a neighbor block, instead manage all neighbors with separate resource.
```
resource "fortios_router_bgp" "fgt_bgp_settings" {
  router_id = 10.1.1.1
  as = 65000
}

resource "fortios_routerbgp_neighbor" "fgt_bgp_neighbor_s1" {
  ip                   = "10.10.10.10"
  remote_as            = 65001
  soft_reconfiguration = "enable"
}

resource "fortios_routerbgp_neighbor" "fgt_bgp_neighbor_s2" {
  ip                   = "10.10.11.10"
  remote_as            = 65001
  soft_reconfiguration = "enable"
}
```
Module down the track adds VPN interface, neighbor etc. Then could be cleanly removed by removing the module:
```
resource "fortios_routerbgp_neighbor" "fgt_bgp" {
  ip                   = data.azurerm_virtual_network_gateway.vnet_gw.bgp_settings[0]["peering_address"]
  remote_as            = data.azurerm_virtual_network_gateway.vnet_gw.bgp_settings[0]["asn"] 
  soft_reconfiguration = "enable"
  update_source        = fortios_vpnipsec_phase1interface.vpn_p1.name
}
```
This example would configure BGP settings and explicitly define no neighbors (and remove existing ones upon apply):
```
resource "fortios_router_bgp" "fgt_bgp_settings" {
  router_id = 10.1.1.1
  as = 65000
  neighbor = {}
}
```
Then this example would be _incorrect_ usage and result in the issues you outlined in point 4:
```
resource "fortios_router_bgp" "fgt_bgp" {
  router_id = 10.1.1.1
  as = 65000
  neighbor {
    ip                   = "10.10.10.10"
    remote_as            = 65001
    soft_reconfiguration = "enable"
  }
}
resource "fortios_routerbgp_neighbor" "fgt_bgp_neighbor_s2" {
  ip                   = "10.10.11.10"
  remote_as            = 65001
  soft_reconfiguration = "enable"
}
```

Yes I have been using the dynamic blocks you mentioned in 5 but this can be hard to read, especially when I am trying to get colleagues on board the IAC train.

Obviously asking for all 400 sub tables to have their own resources would be a bit much but I don't think it would be necessary for most resources.

Is my thinking correct or will there still be conflicts/inconsistencies purely due to how the FortiOS API/Terraform works?

**Comment by frankshen01** [comment]
New resource fortios_routerbgp_neighborhas been supported and released. Support router bgp to clear its sub table when it is destroyed. please see the latest version(v1.7.0): https://registry.terraform.io/providers/fortinetdev/fortios/latest , https://registry.terraform.io/providers/fortinetdev/fortios/latest/docs/resources/fortios_routerbgp_neighbor and https://registry.terraform.io/providers/fortinetdev/fortios/latest/docs/resources/fortios_router_bgp

**Validation:**
```
# cat main.tf
provider "fortios" {
  hostname = "192.168.52.177"
  insecure = "true"
}

resource "fortios_router_bgp" "fgt_bgp" {
  router_id = "10.1.1.1"
  as        = 65000
}

resource "fortios_routerbgp_neighbor" "s1" {
  ip                   = "123.1.1.1"
  remote_as            = 65003
  soft_reconfiguration = "enable"

  depends_on = [
    fortios_router_bgp.fgt_bgp,
  ]
}

# terraform apply

................

mysho (bgp) # show
config router bgp
    set as 65000
    set router-id 10.1.1.1
    config neighbor
        edit "123.1.1.1"
            set soft-reconfiguration enable
            set remote-as 65003
        next
    end
    config redistribute "connected"
    end
    config redistribute "rip"
    end
    config redistribute "ospf"
    end
    config redistribute "static"
    end
    config redistribute "isis"
    end
    config redistribute6 "connected"
    end
    config redistribute6 "rip"
    end
    config redistribute6 "ospf"
    end
    config redistribute6 "static"
    end
    config redistribute6 "isis"
    end
end

# terraform destroy

................

mysho (bgp) # show
config router bgp
    set as 65000
    set router-id 10.1.1.1
    config redistribute "connected"
    end
    config redistribute "rip"
    end
    config redistribute "ospf"
    end
    config redistribute "static"
    end
    config redistribute "isis"
    end
    config redistribute6 "connected"
    end
    config redistribute6 "rip"
    end
    config redistribute6 "ospf"
    end
    config redistribute6 "static"
    end
    config redistribute6 "isis"
    end
end

```
Thanks!

**Comment by poroping** [comment]
Awesome, tested the neighbor resource and it works as requested, thanks!

**Comment by poroping** [comment]
@frankshen01 I know you're busy but when you have the chance can you implement similar resources for:
bgp_network
bgp_network6
ospf_interface
ospf_network
ospf6_interface
ospf6_network

Use case is we are MSSP with hosted VDOM offerings. These resources would be used as part of our internal infrastructure state during VDOM provisioning whilst still allowing additional resources to be easily created within the customer configuration state where needed. eg. customer has s2s vpn running OSPF to connect to the VDOM.

**Comment by frankshen01** [comment]
> @frankshen01 I know you're busy but when you have the chance can you implement similar resources for:
> bgp_network
> bgp_network6
> ospf_interface
> ospf_network
> ospf6_interface
> ospf6_network
> 
> Use case is we are MSSP with hosted VDOM offerings. These resources would be used as part of our internal infrastructure state during VDOM provisioning whilst still allowing additional resources to be easily created within the customer configuration state where needed. eg. customer has s2s vpn running OSPF to connect to the VDOM.

@poroping, thanks for the feature requests.  Thank you for your understanding, it is very precious! I have some stressful delayed tasks whose deadlines must be completed by the end of March. I am afraid that these new features will have to be postponed until April to be fulfilled. If there is a chance during this period, I will try my best to give priority to these features. If it is urgent on your side, you can directly contact the SE serving you to push our product department to adjust your needs to a higher priority. Thank you!

**Comment by poroping** [comment]
No worries, @frankshen01  I'll reach out to my SE.

**Comment by frankshen01** [comment]
@poroping 
```
bgp_network
bgp_network6
ospf_interface
ospf_network
ospf6_interface
```
Supported and released, please see the latest version(v1.13.0): https://registry.terraform.io/providers/fortinetdev/fortios/latest   for details. BTW: I don’t seem to find the ospf6 network you mentioned in the API. Thanks!
